File: lab5/Dictionary.c
//-----------------------------------------------------------------------------
//  Ravi Gupta
//  radgupta
//  CMPS 12B/M
//  5/18/2019
//  Dictionary.c
//  Implementation file for Dictionary ADT
//-----------------------------------------------------------------------------

#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include<assert.h>
#include"Dictionary.h"


// private types --------------------------------------------------------------

// NodeObj
typedef struct NodeObj{
   char* key;
   char* value; 
   struct NodeObj* next;
} NodeObj;

// Node
typedef NodeObj* Node;

// newNode()
// constructor of the Node type
Node newNode(char* x, char* y) {
   Node N = malloc(sizeof(NodeObj));
   assert(N!=NULL);
   N->key = x;
   N->value = y;
   N->next = NULL;
   return(N);
}

// freeNode()
// destructor for the Node type
void freeNode(Node* pN){
   if( pN!=NULL && *pN!=NULL ){
      free(*pN);
      *pN = NULL;
   }
}
// DictionaryObj
typedef struct DictionaryObj{
   Node head;
   int numPairs;
} DictionaryObj;

// newDictionary()
// constructor for the Dictionary type
Dictionary newDictionary(void){
   Dictionary D = malloc(sizeof(DictionaryObj));
   assert(D!=NULL);
   D->head = NULL;
   D->numPairs = 0;
   return D;
}

// freeDictionary()
// destructor for the Dictionary type
void freeDictionary(Dictionary* pD){
   if( pD!=NULL && *pD!=NULL ){
      makeEmpty(*pD);
      free(*pD);
      *pD = NULL;
   }
}

// findKey()
// returns the Node containing key k or returns
// NULL if no such Node exists
Node findKey(Dictionary D, char* key){
   Node N = D->head;
   while( N != NULL) {
      if ( strcmp(key, N->key) == 0 ){ 
         return N;
      }
      N = N->next;
   }
   return NULL;
}

// isEmpty()
// returns 1 (true) if D is empty, 0 (false) otherwise
// pre: none
int isEmpty(Dictionary D){
   if( D==NULL ){
      fprintf(stderr, 
         "Dictionary Error: calling isEmpty() on NULL Dictionary reference\n");
      exit(EXIT_FAILURE);
   }
   if( D->numPairs==0 ) return 1;
   return 0;
}

// size()
// returns the number of (key, value) pairs in D
// pre: none
int size(Dictionary D){
   if( D==NULL ){
      fprintf(stderr, 
         "Dictionary Error: calling size() on NULL Dictionary reference\n");
      exit(EXIT_FAILURE);
   }
   return(D->numPairs);
}

// lookup()
// returns the value v such that (k, v) is in D, or returns NULL if no 
// such value v exists.
// pre: none
char* lookup(Dictionary D, char* k){
   Node N;
   if( D==NULL ){
      fprintf(stderr, 
         "Dictionary Error: calling lookup() on NULL Dictionary reference\n");
      exit(EXIT_FAILURE);
   }
   N = findKey(D, k);
   if ( N==NULL) return NULL;
   return N->value;
}

// insert()
// inserts new (key,value) pair into D
// pre: lookup(D, k)==NULL
void insert(Dictionary D, char* k, char* v){
   Node N, R;
   if( D==NULL ){
      fprintf(stderr, 
         "Dictionary Error: calling insert() on NULL Dictionary reference\n");
      exit(EXIT_FAILURE);
   }
   if( lookup(D, k)!=NULL ){
      fprintf(stderr, 
         "Dictionary Error: cannot insert() duplicate key: \"%s\"\n", k);
      exit(EXIT_FAILURE);
   }
   N = newNode(k, v);
   if( D->head == NULL){
      D->head = N;
   }
   else{
      R = D->head;
      while( R->next != NULL ){
         R = R->next;
      }
      R->next = N;
   }
   D->numPairs++;
}

// delete()
// deletes pair with the key k
// pre: lookup(D, k)!=NULL  
void delete(Dictionary D, char* k){
   Node N, R;
   if( D==NULL ){
      fprintf(stderr, 
         "Dictionary Error: calling delete() on NULL Dictionary reference\n");
      exit(EXIT_FAILURE);
   }
   if( lookup(D, k)==NULL ){
      fprintf(stderr, 
         "Dictionary Error: cannot delete() non-existent key: \"%s\"\n", k);
      exit(EXIT_FAILURE);
   }
   N = findKey(D, k);
   if(N == D->head){
      D->head = N->next;
      freeNode(&N);
   } 
   else{
      R = D->head;
      while( R->next != NULL ){
          if( strcmp(N->key, R->next->key)==0 ){
              break;
          }
          R = R->next;
      }
      R->next = N->next;
      freeNode(&N);
   }
   D->numPairs--; 
}

// makeEmpty()
// re-sets D to the empty state.
// pre: none
void makeEmpty(Dictionary D){
   Node R, N = D->head;
   while( N != NULL ){
      R = N;
      N = N->next;
      freeNode(&R);
      D->numPairs--;
   }
}
// printDictionary()
// pre: none
// prints a text representation of D to the file pointed to by out
void printDictionary(FILE* out, Dictionary D){
   if( D==NULL ){
      fprintf(stderr, 
         "Dictionary Error: calling printDictionary() on NULL"
         " Dictionary reference\n");
      exit(EXIT_FAILURE);
   }
   Node N = D->head;
   while( N != NULL ){
      fprintf(out, "%s %s\n", N->key, N->value);      
      N = N->next;
   } 
}

File: pa3/Dictionary.java
//-----------------------------------------------------------------------------
// Ravi Gupta
// radgupta
// CMPS12B/M
// 5/4/19
// Dictionary.java
// Linked List implementation of the Dictionary ADT
//-----------------------------------------------------------------------------

public class Dictionary implements DictionaryInterface {

   // private inner Node class
   private class Node {
      String key;
      String value;
      Node next;
    
      Node(String x, String y){
         key = x;
         value = y;
         next = null;
      }
   }

   // Fields for the Dictionary class
   private Node head;     // reference to first Node in List
   private int numValues;  // number of items in this Dictionary
   // Dictionary()
   // constructor for the Dictionary class
   public Dictionary(){
      head = null;
      numValues = 0;
   }
   // private helper function -------------------------------------------------
   
   // findKey()
   // returns a reference to the Node at position index in this Dictionary
   private Node findKey(String x){
      for (Node N = head; N != null; N = N.next) {
         if ( x.equals(N.key)){
            return N;
         }
      }
      return null;   
   }
   
   // ADT operations ----------------------------------------------------------
   
   // isEmpty()
   // pre: none
   // post: returns true if this Dictionary is empty, false otherwise
   public boolean isEmpty(){
      return(numValues == 0);
   }
  
   // size()
   // pre: none
   // post: returns the number of entries in the Dictionary
   public int size() {
      return numValues;
   }
   
   // lookup()
   // pre: none
   // returns value associated key, or null reference if no such key exists
   public String lookup(String key){
      Node N = findKey(key);
      if ( N == null ){
         return null;
      } 
      return N.value; 
   }
 
   // insert()
   // inserts new (key,value) pair into this Dictionary
   // pre: lookup(key)==null
   public void insert(String key, String value) throws DuplicateKeyException{
      if( lookup(key) == null ){
         Node N = new Node(key, value);
         if( head == null){
            N.next = head;
            head = N;
         }
         else{
            Node R = head;
            while( R.next != null ){
              R = R.next;
            }           
            R.next = N;
         }
         numValues++;
      }
      else{   
         throw new DuplicateKeyException("cannot insert duplicate keys");
      }
   }
   // delete()
   // deletes pair with the given key
   // pre: lookup(key)!=null
   public void delete(String key) throws KeyNotFoundException{
      if( lookup(key) != null ){
         Node N = findKey(key);
         if(N == head){
            head = N.next;
         }
         else{
            Node R = head;
            while( R.next != null ){
               if( N.key.equals(R.next.key)){
                  break;
               }
               R = R.next;
            }
            R.next = N.next;
            N.next = null;
         }
         numValues--;
      }
      else{
         throw new KeyNotFoundException("cannot delete non-existent key");
      }
   } 
   // makeEmpty()
   // pre: none
   public void makeEmpty(){
      head = null;
      numValues = 0;
   }
   // toString()
   // returns a String representation of this Dictionary
   // overrides Object's toString() method
   // pre: none
   public String toString(){
      StringBuffer sb = new StringBuffer();
      Node N = head;

      for( ; N!=null; N=N.next){
         sb.append(N.key).append(" ").append(N.value).append("\n");
      }
      return new String(sb);      
   }
}
